hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/AbstractHandler.java

```diff
@@ -98,11 +101,33 @@ final void unregisterOp(int operation) throws IOException {
         selectionKey.interestOps(selectionKey.interestOps() & ~operation);
     }
 
+    @SuppressWarnings({"checkstyle:npathcomplexity"})
     @Override
     public void onFailure(Throwable e) {
         if (e instanceof OutOfMemoryError) {
             ioService.onOutOfMemory((OutOfMemoryError) e);
         }
+
+        // if our selection key was cancelled due to replacement of the selector
+        // (see NonBlockingIOThread#selectWorkaround), then reschedule this handler for execution
+        if (e instanceof CancelledKeyException && selectionKey != null
+                && selectionKey.selector() != ioThread.getSelector()) {
+            // the selector was replaced, so obtain a new key for the new selector
+            SelectableChannel ch = selectionKey.channel();
+            int ops = selectionKey.interestOps();
+            Object att = selectionKey.attachment();
+
+            // register the channel with the new selector now
+            try {
+                SelectionKey sk = ch.register(ioThread.getSelector(), ops, att);
+                this.selectionKey = sk;
+            } catch (ClosedChannelException ignore) {
+                EmptyStatement.ignore(ignore);
```

Comment:
Instead of returning, I would continue with the original flow of the exception. So closing the connection.